### PR TITLE
Add option --max-columns to truncate long lines

### LINKED ISSUE #129: maximum line length
I'd like `ripgrep` to have the ability to either hide or trim lines that are very long. Some lines take up my entire screen and are borderline useless to look at. It's possible that finding an intelligent way to shorten them would be best, since my guess is that the actual matched text is much smaller than the full line. However, this is harder to implement.

I don't think this should be enabled by default. It seems a little surprising for `ripgrep` to hide lines like that. In general, I like the work flow of, "run a search, see huge lines, confirm that I don't care about them and run ripgrep again with an option to hide them." It may however be plausible to enable this limit if results are being dumped to a terminal (we already enable colors, line numbers and file headings).

### PR DESCRIPTION
**Pull request by RalfJung**
This implements what we discussed in <https://github.com/BurntSushi/ripgrep/issues/129#issuecomment-276946474>.

One open question from my side: how bad is it that by always using `CountingReplacer`, we lose the optimization offered by returning something from `Replacer::no_expansion`? In principle, we could use `CountingReplacer` only if `max_columns` is `Some`, therefore only incurring a performance cost if `--max-columns` is set. Is that worth the additional effort?

TODO: Add some tests.

Fixes #129.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -454,6 +457,10 @@ lazy_static! {
              "Show results with every match on its own line, including \
               line numbers and column numbers. With this option, a line with \
               more than one match will be printed more than once.");
+        doc!(h, "max-columns",
+             "Don't print lines wider than NUM lines.",
+             "Don't print lines wider than NUM lines.  Longer lines are \
```
Comment (BurntSushi) [inline src/app.rs]: We should try to keep a consistent style. One space after periods please. :-)

File: src/printer.rs [hunk 2]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
             self.write(b":");
         }
         if self.replace.is_some() {
-            let line = re.replace_all(
-                &buf[start..end], &**self.replace.as_ref().unwrap());
+            let mut count = 0;
+            let line = {
+                let replacer = CountingReplacer { count: &mut count, replace : &**self.replace.as_ref().unwrap() };
```
Comment (BurntSushi) [inline src/printer.rs]: I think we should write code that is consistent with the style of the surrounding code. Could you please adhere to a 79 column limit (inclusive)? Thanks.

File: src/printer.rs [hunk 3]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
+                re.replace_all(&buf[start..end], replacer)
+            };
+            if let Some(max_columns) = self.max_columns {
+                if line.len() > max_columns {
```
Comment (BurntSushi) [inline src/printer.rs]: `if self.max_columns.map_or(false, |m| line.len() > m)`

File: src/printer.rs [hunk 4]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
+                    let _ = self.wtr.set_color(self.colors.line());
+                    self.write(format!(" [Omitted long line with {} replacements]", count).as_bytes());
```
Comment (BurntSushi) [inline src/printer.rs]: I don't think there should be a leading space here.

File: src/printer.rs [hunk 5]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
-            let line = re.replace_all(
-                &buf[start..end], &**self.replace.as_ref().unwrap());
+            let mut count = 0;
+            let line = {
+                let replacer = CountingReplacer { count: &mut count, replace : &**self.replace.as_ref().unwrap() };
+                re.replace_all(&buf[start..end], replacer)
+            };
+            if let Some(max_columns) = self.max_columns {
+                if line.len() > max_columns {
+                    let _ = self.wtr.set_color(self.colors.line());
```
Comment (BurntSushi) [inline src/printer.rs]: I think I would rather see this use the match color instead of the line number color.

File: src/printer.rs [hunk 6]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
-        }
-        if buf[start..end].last() != Some(&self.eol) {
-            self.write_eol();
+            // write_matched_line guarantees to write a newline.
         }
     }
 
     fn write_matched_line(&mut self, re: &Regex, buf: &[u8]) {
+        if let Some(max_columns) = self.max_columns {
+            if buf.len() > max_columns {
```
Comment (BurntSushi) [inline src/printer.rs]: `if self.max_columns.map_or(false, |m| buf.len() > m)`

File: src/printer.rs [hunk 7]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
+                let count = re.find_iter(buf).count();
+                let _ = self.wtr.set_color(self.colors.line());
+                self.write(format!(" [Omitted long line with {} matches]", count).as_bytes());
```
Comment (BurntSushi) [inline src/printer.rs]: I don't think we need the leading space here.

File: src/printer.rs [hunk 8]
```diff
@@ -263,31 +286,57 @@ impl<W: WriteColor> Printer<W> {
...
-            self.write_eol();
+            // write_matched_line guarantees to write a newline.
         }
     }
 
     fn write_matched_line(&mut self, re: &Regex, buf: &[u8]) {
+        if let Some(max_columns) = self.max_columns {
+            if buf.len() > max_columns {
+                let count = re.find_iter(buf).count();
+                let _ = self.wtr.set_color(self.colors.line());
```
Comment (BurntSushi) [inline src/printer.rs]: "match" color instead of "line" color.

File: src/printer.rs [hunk 9]
```diff
@@ -312,6 +361,15 @@ impl<W: WriteColor> Printer<W> {
         if let Some(line_number) = line_number {
             self.line_number(line_number, b'-');
         }
+        if let Some(max_columns) = self.max_columns {
+            if end - start > max_columns {
```
Comment (BurntSushi) [inline src/printer.rs]: `if self.max_columns.map_or(false, |m| end - start > m)`

File: src/printer.rs [hunk 10]
```diff
@@ -312,6 +361,15 @@ impl<W: WriteColor> Printer<W> {
...
+                let _ = self.wtr.set_color(self.colors.line());
+                self.write(format!(" [Omitted long context line]").as_bytes());
```
Comment (BurntSushi) [inline src/printer.rs]: Drop leading space.

**Comment by BurntSushi** [comment]
@RalfJung Thank you for working on this. :-)

**Comment by BurntSushi** [comment]
I'd like to push a new release out soon and you got this PR so close to completion that I picked it up in #406. Thanks so much!

**Comment by RalfJung** [comment]
Ah, so sorry for not coming back to this earlier. Busy times. I was hoping to do something this week. Thanks for completing this!

**Pull request closed without merging**